zcash_client_backend/src/data_api/wallet/input_selection.rs

```diff
@@ -0,0 +1,398 @@
+        &self.transaction_request
+    }
+    pub fn transparent_inputs(&self) -> &[TransparentInput] {
+        &self.transparent_inputs
+    }
+    pub fn sapling_inputs(&self) -> &[SpendableNote<NoteRef>] {
+        &self.sapling_inputs
+    }
+    pub fn balance(&self) -> &TransactionBalance {
+        &self.balance
+    }
+    pub fn fee_rule(&self) -> &FeeRuleT {
+        &self.fee_rule
+    }
+}
+
+/// A strategy for selecting transaction inputs and proposing transaction outputs.
+///
+/// Inputs returned should include only economically useful inputs with economic usefulness
+/// determined according to the fee estimator; that is, do not return inputs that cause fees to
+/// increase by an amount greater than the value of the input.
+pub trait InputSelector {
+    /// The type of errors that may be generated in input selection
+    type Error;
+    /// The type of data source that the input selector expects to access
+    /// to obtain input notes and UTXOs. This associated type permits input
+    /// selectors that may use specialized knowledge of the internals of a
+    /// particular backing data store, if the generic API of `WalletRead`
+    /// does not provide sufficiently fine-grained operations for a particular
+    /// backing store to optimally perform input selection.
+    type DataSource: WalletRead;
+    /// The type of the fee rule that this input selector uses when computing fees.
+    type FeeRule: FeeRule;
+
+    /// Returns a list of spendable inputs sufficient to cover the amount required to satisfy the
+    /// given [`TransactionRequest`], with fees as estimated by the provided [`FeeEstimator`]
+    ///
+    /// Implementations of this method should return inputs sufficient to satisfy the given
+    /// transaction request using a best-effort strategy to preserve user privacy, as follows:
+    /// * If it is possible to satisfy the specified transaction request by creating
+    ///   a fully-shielded transaction without requiring value to cross pool boundaries,
+    ///   return the inputs necessary to construct such a transaction; otherwise
+    /// * If it is possible to satisfy the transaction request by creating a fully-
+    ///   shielded transaction with some amounts crossing between shielded pools,
+    ///   return the inputs necessary.
+    /// If insufficient funds are available to satisfy the required outputs for the shielding
+    /// request, this operation must fail and return [`InputSelectorError::InsufficientFunds`].
+    #[allow(clippy::too_many_arguments)]
+    #[allow(clippy::type_complexity)]
+    fn propose_transaction<ParamsT>(
+        &self,
+        params: &ParamsT,
+        wallet_db: &Self::DataSource,
+        account: AccountId,
+        anchor_height: BlockHeight,
+        target_height: BlockHeight,
+        transaction_request: TransactionRequest,
+    ) -> Result<
+        Proposal<
+            Self::FeeRule,
+            std::convert::Infallible,
```

Comment:
Is this the right type to use here, or should this use `()`? Is it valid to construct an empty vector of uninhabitable types (clearly Rust lets you)? What happens if the caller attempts to iterate over them?